src/adapter/src/client.rs

```diff
@@ -532,9 +548,25 @@ impl SessionClient {
                     results.push(SimpleResult::err(
                         "executing statements of this type is unsupported via this API",
                     ));
+                    break;
                 }
             };
         }
+
+        let action = if matches!(results.last(), Some(SimpleResult::Err { .. }))
+            || matches!(
+                self.session().transaction(),
+                TransactionStatus::InTransaction(..),
+            ) {
+            // If the statements errored, or the last statement did not close an
+            // open explicit transaction, rollback.
+            EndTransactionAction::Rollback
+        } else {
+            // If this is a pointless commit, no harm
+            EndTransactionAction::Commit
+        };
+        self.end_transaction(action).await?;
```

Comment:
This is bugged on `main` in `simple_execute`, e.g. `insert into t values (1)` is treated as an explicit transaction because it contains only one message. When the statements are done processing, nothing performs the commit and the write is lost.

Additionally, we have to check for the last result being an error because we perform checks outside of SQL in this API, e.g. if in an implicit transaction and the final statement contains parameters, the final result will be an error and the transaction must be rolled back and not committed.

I chose this approach rather than using `fail_transaction` because the docs for `TransactionStatus::Failed`  say:

>  In a failed transaction that was started explicitly (i.e., previously 
> InTransaction). We do not use Failed for implicit transactions because
> those cleanup after themselves. Matches `TBLOCK_ABORT`.

which indicates to me that it's is inappropriate to mark an implicit transaction as failed.

Reply:
The docs for `TransactionStatus::Failed` are wrong and should be fixed. Implicit and explicit txns both use Failed. I agree something needs to commit at the end. We should use the same fail machinery as protocol.rs

My current thoughts on the best way to do txns + params are to use exactly the same stuff as the extended protocol, with a `Sync` at the end. That will correctly handle txns, params, aborted txns, auto commit if still implicit at the end, and work nicely with a web api that can send arrays of single statement query strings with params.

Reply:
Ah, didn't know that the docs were wrong. Fixed.